### PR TITLE
(feat) add support for `musl` using `cross-rs`

### PR DESCRIPTION
**Pull request by jprendes**
This PR adds support for building and testing against `musl` targets.

A summary of changes:
* Add musl targets to `Cross.toml` and the corresponding `Dockerfile.musl`
* Install dependencies in `Dockerfile.gnu` so that we can run tests using cross.
* Add a `cargo.sh` script that wraps `cargo` calls, using either `cargo` or `cross` based on the target.
* Fix `mock.rs` so that it builds with `musl` targets
* Use the `cargo.sh` script in the `features_test.sh` and `justfile`
* Remove now redundant `musl_test.sh`.
* Update CI to install `cross`, use `cargo.sh`, and force using `cross` for all targets (including host target).
* Update CI to relese `musl` binaries as well as `gnu` binaries.

Example of a release generated from this PR: https://github.com/jprendes/youki/releases/tag/v0.4.1

Note: This PR builds on top of #2541. I will rebase once that is merged.

### CODE REVIEW

**Comment by utam0k** [comment]
Sorry, I won't be able to find the time this week, so I'll look at it in a few days 🙏

File: scripts/cargo.sh
Comment (adrianalin) [inline scripts/cargo.sh:26]: Isn't this overwritten by
https://github.com/containers/youki/pull/2536/files#diff-3d64996115d0ac3b74e36f57ce6b00d6b4982614c41b1ed39eab8ef43b2e9a2bR38 ?

File: scripts/cargo.sh
Comment (adrianalin) [inline scripts/cargo.sh:34]: No need to add new dependency (jq), you can use 
```suggestion
ROOT=$(git rev-parse --show-toplevel)
```

File: scripts/cargo.sh
Comment (adrianalin) [inline scripts/cargo.sh:1]:
```suggestion
#!/usr/bin/env bash
set -euo pipefail
```

File: scripts/cargo.sh
Comment (jprendes) [reply scripts/cargo.sh:27]: Thanks, good catch :-)

**Comment by jprendes** [comment]
Reabsed.
@utam0k @adrianalin @cuisongliu PTAL

An example release generated by this PR is:
https://github.com/jprendes/youki/releases/tag/v0.3.2

**Comment by YJDoc2** [comment]
Self-reminder : need to update  merge CI requirements to correct CI names after this is merged.
(The three pending CI seems to be because of renaming of the workflows, so should be ok.)

**Review by utam0k — APPROVED** [review]
May I ask a favor of the documentation for developers in another PR?
https://containers.github.io/youki/developer/introduction.html

**Comment by utam0k** [comment]
Thanks!! @jprendes

**Comment by utam0k** [comment]
> Self-reminder : need to update merge CI requirements to correct CI names after this is merged. (The three pending CI seems to be because of renaming of the workflows, so should be ok.)

Done ✅

**Comment by YJDoc2** [comment]
Thanks @utam0k 
Is this ready to merge? The CI checks which are updated in this are keeping other PRs from passing...
I can revert the CI check requirements temp until this is merged, or as we don't have anything urgent, just wait out for this.
Let me know.

File: .github/workflows/main.yml
Comment (YJDoc2) [inline .github/workflows/main.yml:59]: If I'm right, then I don't think we need matrix for lint checking, the `just lint` target should checks for all targets and features. Would that be affected by arch or libc type?

File: Cross.toml
Comment (YJDoc2) [inline Cross.toml:2]: I think we should keep the default target to gnu, rather than musl unless we have some strong reason not to...

File: scripts/build.sh
Comment (YJDoc2) [inline scripts/build.sh:1]: If I'm correct, this would still use the normal cargo for building unless we specify cross, right? I'd prefer not needing cross or other related deps for local development unless needed in development.

**Review by YJDoc2 — APPROVED** [review]
LGTM :+1:

I have some nitpicks / questions, but nothing that would block merging, so as @utam0k has already approved, I'll wait sometime and merge this.

@jprendes thanks a lot for this, this is really amazing work! Take your time for replying to the comments :+1:

File: scripts/build.sh
Comment (jprendes) [reply scripts/build.sh:1]: Thats correct. It defaults to cargo when building for the host triple, and to cross when building for another triple.

Cross can be forced with the env var `CARGO=cross` or with the `-x` CLI argument.

File: .github/workflows/main.yml
Comment (jprendes) [reply .github/workflows/main.yml:59]: The arch could make a difference with clippy when conditional compilation is used.

This also applies for dependencies using conditional compilation. An example of this is in `mock.rs` in this PR, where libc doesn't expose `__rlimit_resource_t` when building for musl.

File: Cross.toml
Comment (jprendes) [reply Cross.toml:2]: This only impacts when cross is called without specifying a target. The cargo.sh script always specifies a target, so it doesn't use the default.

I'll change it to gnu for consistency.

**Comment by utam0k** [comment]
Ops sorry I'll merge it

**Comment by adrianalin** [comment]
Sorry, I could not find where documentation was updated, @jprendes please provide link. Thank you!

**Comment by jprendes** [comment]
> Sorry, I could not find where documentation was updated (related to musl build)

I still need to create a PR. I'll do that as soon as I have some bandwidth again.

**Pull request merged**
